test_agent/rust/src/testagent.rs

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (c) 2024 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use async_trait::async_trait;
+use log::error;
+use serde_json::Value;
+use up_rust::{Data, UCode, UListener};
+use up_rust::{UMessage, UStatus, UTransport};
+
+use std::io::{Read, Write};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use serde::Serialize;
+
+use crate::utils::{convert_json_to_jsonstring, WrapperUMessage, WrapperUUri};
+use crate::{constants, utils, TcpStreamSync, UTransportSocket};
+
+use self::utils::sanitize_input_string;
+
+#[derive(Serialize)]
+pub struct JsonResponseData {
+    data: HashMap<String, String>,
+    action: String,
+    ue: String,
+    test_id: String,
+}
+#[derive(Clone)]
+pub struct SocketTestAgent {
+    utransport: UTransportSocket,
+    clientsocket: Arc<Mutex<TcpStreamSync>>,
+    clientsocket_to_tm: Arc<Mutex<TcpStreamSync>>,
+}
+
+#[async_trait]
+impl UListener for SocketTestAgent {
+    async fn on_receive(&self, msg: UMessage) {
+        dbg!("OnReceive called");
+
+        let data_payload = match &msg.payload.data {
+            Some(data) => {
+                // Now we have access to the Data enum
+                match data {
+                    Data::Reference(reference) => reference.to_string(),
+                    Data::Value(value) => {
+                        let value_str = String::from_utf8_lossy(value);
+                        value_str.to_string()
+                    }
+                    _ => "none".into(),
+                }
+            }
+            None => {
+                dbg!("No data available");
+                "none".into()
+            }
+        };
+
+        let mut value: HashMap<String, String> = HashMap::new();
```

Comment:
IMHO it's worthwhile to revisit the suggestion I gave [here](https://github.com/eclipse-uprotocol/up-tck/pull/23#discussion_r1565931126) to improve code readability and style.

Generally speaking in this case you need not use mutable variables, but rather construct the JsonResponseData out of elements which are already all set.